### PR TITLE
[Shep 123] - Enabling Sentry for React frontend of ad-ops-dashboard

### PR DESCRIPTION
**Pull request by luc-lisi**
## References

[SHEP-123](https://mozilla-hub.atlassian.net/browse/SHEP-123)

## Problem Statement

In this ticket we aim to enable sentry for our [React frontend](https://docs.sentry.io/platforms/javascript/guides/react/). This will allow better logging, visibility, and handling of browser-side errors that might occur for the user.

## Proposed Changes

In this PR, we've followed the above docs _mostly_ but had to make a few modifications in order to have things work with `react-router`. Information regarding this integration can be found [here](https://docs.sentry.io/platforms/javascript/guides/react/features/react-router/)

At a high level, the following changes were made in this PR:

1. A new `instruments.tsx` file which contains the `Sentry.init()` as well as configuration options, and the definition and export of a `sentryCreateBrowserRouter` wrapper for the `react-router` package.
2. We've modified the `App.tsx` file to use this new `sentryCreateBrowserRouter` wrapper for our routes. This operates the same as it did previously, but allows us to handle error boundaries without react-router "eating" the errors before Sentry can see them.
3. We've modified the `vite-config.ts` file in order to surface the `NODE_ENV` environment variable. This is a really useful env variable that allows us to lock certain features behind dev only mode. (dev mode is only enabled when running the app through `npm run dev`. Productions builds will have this set to NODE_ENV='production')
4. In order to keep Sentry from logging errors in local dev, we've added a check for this env variable and an override in our Sentry Options to log simply console log the Sentry event rather than push it to Sentry.
5. A new "DefaultError" component was created as a placeholder for any default error page we might want. For now I'm just including something so we can test the Sentry error boundaries. 

In addition, I've added an extra route `/debug` which is only accessible in dev mode. In any production build, this will throw a 404. I've done this in the past for my frontends and it's been really useful to add different buttons that can force some state easily in the application. For example, here we are using it to test the sentry error boundaries and sentry logging by having buttons that force certain failure states. Let me know if folks like this.

## Verification Steps

- [ ] Install all new dependencies
- [ ] run the app using `npm run dev` or whatever other dev entry point you'd like
- [ ] navigate to `/debug` (Verify this is does not 404)
- [ ] Open the browser console and verify a sentry event is created when the first button `Throw caught sentry error` is hit.
- [ ] Press `Break useEffect` and verify a sentry event is logged, and the page gracefully fails over to the DefaultError component.
- [ ] Go back to `/debug` and verify the last button `Break Page` does the same. (These last two steps are two different failure modes. One where a `useEffect` call fails and another where the `JSX` enters an invalid state)
- [ ] Finally create a local production build of the app using `npm run build` followed by `npm run start` (you may have to close any docker containers you have running)
- [ ] Re-open the fronted in the browser and check `/debug` throws a 404.

## Visuals

If applicable, provide screenshots or videos here with the expected new behavior. If not, please remove this section.

### CODE REVIEW

File: ad-ops-dashboard/src/instrument.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    Sentry.browserTracingIntegration(),
+  ],
+  environment: process.env.NODE_ENV,
+  // Set tracesSampleRate to 1.0 to capture 100%
+  // of transactions for tracing.
+  tracesSampleRate: 1.0,
+}
+
+// Log sentry events to the console, not sentry when on local dev
+if (isLocal) {
```
Comment (mashalifshin) [inline ad-ops-dashboard/src/instrument.tsx:35]: I wonder if we might avoid this clause entirely if we don't set the Sentry environment variables at all for local development. 

This seems to be the approach taken on [the Django side](https://github.com/mozilla-services/consvc-shepherd/blob/main/consvc_shepherd/settings.py#L235). Also looks like they are using `SENTRY_MODE` which can configure sentry to log out detailed debug info, and may be sufficient here? 

Or maybe we can use `SENTRY_MODE` to make the decision on whether to log errors to the console here, if that info about what we're sending to Sentry is useful beyond the usual stack trace we'd get on an error

File: ad-ops-dashboard/src/components/Errors/DefaultError.tsx [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import { useEffect } from "react";
+
+export function DefaultError() {
+  const error = useRouteError() as Error;
+
+  useEffect(() => {
+    Sentry.captureException(error);
+  }, [error]);
+
+  return (
```
Comment (mashalifshin) [inline ad-ops-dashboard/src/components/Errors/DefaultError.tsx:12]: I noticed that when we are navigated to the error page, the path in the url bar isn't updated to `/error` and just stays at `/debug`. Wondering if this is the intended behavior since it's an error page? I see that when I navigate to the Campaign page, the path gets updated in that case...

File: ad-ops-dashboard/src/instrument.tsx (hunk 1 above)
Comment (mashalifshin) [reply ad-ops-dashboard/src/instrument.tsx:35]: If we do want to verify that the Sentry integration is behaving as expected, is there a way we could do that with a unit test?

**Review by mashalifshin — CHANGES_REQUESTED** [review]
Loving that new debug page, as well as the new error page. I took it for a spin and it's working like a charm! 

I am requesting some changes regarding loading the sentry environment variables, but please lmk if I'm missing stuff here, still learning and building my understanding on all this

**Comment by mashalifshin** [comment]
☑️ Also wanted to mention I enjoyed using the checklist and checking things off when verifying! It made it more fun, and thank you for taking the time to put it together.

File: ad-ops-dashboard/vite.config.ts [hunk 3]
```diff
@@ -1,14 +1,27 @@
...
-  plugins: [react()],
-  server: {
-    host: true,
-    strictPort: true,
-    port: 5173,
-    watch: {
-      usePolling: true,
+export default defineConfig(({ mode }) => {
+  // Load environment variables based on the mode (development, production, etc.)
+  const env = loadEnv(mode, process.cwd());
```
[the remark this answers is not in the document]
Comment (luc-lisi) [reply ad-ops-dashboard/vite.config.ts:6]: Ah so good question! I've used Vite in the past but by no means an expert so take this with a grain of salt. Based on [these docs](https://vite.dev/guide/env-and-mode) my understanding is that `.env` is always read, but _if_ an `.env.production` file is defined it will take priority (assuming we are in this mode). There are also other interesting cases where certain env variables have priority over others. For example, they mention that: `In addition, environment variables that already exist when Vite is executed have the highest priority and will not be overwritten by .env files. For example, when running VITE_SOME_KEY=123 vite build.` So unless I'm totally misunderstanding this, anything set by our docker containers will take priority over everything in any `.env*` files.

So in our case `mode` is effectively ignored but still, I believe, required by the `loadEnv` signature to read in the `.env` file we use locally. Let me know if this makes sense though! I also might be totally missing something 😅 

To your second point though we can totally use the existing `SENTRY_ENV` variable and use that instead of the `NODE_ENV` variable! For now I'll just add it to the `.env.example` file and my local `.env`, but when we do the full integration we will just have to double check it gets bundled in with the build properly.

I'll make that change 😄

File: ad-ops-dashboard/src/components/Errors/DefaultError.tsx (hunk 2 above)
Comment (luc-lisi) [reply ad-ops-dashboard/src/components/Errors/DefaultError.tsx:12]: Ah yeah, this is the intended behavior! At least in this simple case. This is really just meant to be an example component that shows off the simplest use of Sentry's ErrorBoundary. You can add tons of more advanced logic here (like changing the route as a result of a failure) but for the simplest case you might use it just to catch a single component's broken JSX within a page. Otherwise, you end up rendering weird ramblings and logs from react-router or breaking the entire site. In a real implementation of these boundaries though we can totally route us to a completely new page for certain situations! Probably worth adding an entire separate ticket for how exactly we want to handle error boundaries across different pages and components though unless you feel strongly about including that logic here 😄 

Let me go ahead though and change this example in the /debug page as I agree that because the page is really just one big component, it looks functionally the same as if we were to change pages.

Comment (luc-lisi) [reply ad-ops-dashboard/src/components/Errors/DefaultError.tsx:12]: It also doesn't help that I put this inside of the router lol. Good eye though as I totally could have made this clearer!

File: ad-ops-dashboard/src/instrument.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    Sentry.reactRouterV6BrowserTracingIntegration({
+      useEffect: useEffect,
+      useLocation,
+      useNavigationType,
+      createRoutesFromChildren,
+      matchRoutes,
+    }),
+    Sentry.browserTracingIntegration(),
+  ],
+  environment: process.env.NODE_ENV,
```
[the remark this answers is not in the document]
Comment (luc-lisi) [reply ad-ops-dashboard/src/instrument.tsx]: Swapped! Note that we will still have to copy `SENTRY_ENV` to `VITE_SENTRY_ENV` at some point during the build process as Vite wont read in any environment variables without the `VITE` prefix to avoid leaking server secrets to the browser. But that shouldn't be too bad!

File: ad-ops-dashboard/src/instrument.tsx (hunk 1 above)
Comment (luc-lisi) [reply ad-ops-dashboard/src/instrument.tsx:35]: For point 1: Oh good call on the debug flag! Yeah we can totally add that here although in the code you linked it seems like SENTRY_MODE is a string, whereas in the React SDK it's a bool? So maybe I'll just have it mirror `isLocal.` Lmk if you disagree with this though or if I'm missing something. For clarity though, the idea here with this interceptor is just to be able to see in the console the exact event that would be logged to sentry had that error occurred in prod, but not necessarily provide additional debug value for local dev. So ideally for localdev we have both this `debug=true` line and this interceptor I think?

For point 2: I'm not sure about how easy it would be to write full integration tests with sentry, but we for sure can write some unit test to at least see if the local logs emitted from the local Sentry interceptor work as expected! I can add that as part of the Default error page test 😄

File: ad-ops-dashboard/src/pages/Debug/Debug.tsx [hunk 5]
```diff
@@ -0,0 +1,53 @@
+import * as Sentry from "@sentry/react";
```
[the remark this answers is not in the document]
Comment (luc-lisi) [reply ad-ops-dashboard/src/pages/Debug/Debug.tsx:2]: I think in this case I'll definitely add one we can use as an example for testing error boundaries! I'm always a fan of more test coverage haha. I think in general though I'd maybe lean towards minimal unit tests for the debug page directly... but any components or logic the debug page uses (that might also be used outside the page) should 100% have unit tests. Like the debug page feels like a convenient way for us to add UI elements to quickly force state while developing locally, but breaking something in the debug page probably shouldn't block us from shipping a production feature? I also wouldn't want people mistaking adding some functionality to the debug page and adding a test as being equivalent to adding a test for the exact production behavior we want to verify.

Let me know though if that makes sense. And I'll definitely add some unit tests for Sentry!

File: ad-ops-dashboard/src/components/Errors/DefaultError.tsx [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import { useRouteError } from "react-router-dom";
```
[the remark this answers is not in the document]
Comment (luc-lisi) [reply ad-ops-dashboard/src/components/Errors/DefaultError.tsx:1]: Indeed this lil sad path component does!

[the remark this answers is not in the document]
Comment (luc-lisi) [reply ad-ops-dashboard/src/components/Errors/DefaultError.tsx:1]: Added!

File: ad-ops-dashboard/src/pages/Debug/Debug.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (luc-lisi) [reply ad-ops-dashboard/src/pages/Debug/Debug.tsx:2]: Unit tests added to verify the page 404's when not in dev mode. Also refactored the content of the debug into their own components and added some unit tests to those! Figured that made more sense than page-wide unit tests, but let me know if you disagree.

File: ad-ops-dashboard/vite.config.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mashalifshin) [reply ad-ops-dashboard/vite.config.ts:6]: Great, thank you for going in depth and making sure we'll get the right env values here locally and in prod. 

> So in our case mode is effectively ignored but still, I believe, required by the loadEnv signature to read in the .env file we use locally. Let me know if this makes sense though! I also might be totally missing something 😅

Yeah, that makes sense to me, seems like leaving `mode` unused is good for our case since we don't have those per-environment `.env` files. 

> when we do the full integration we will just have to double check it gets bundled in with the build properly.

Yeah I'm glad we dug into this on this PR because I'll now need to make sure those are available at build time! Before we only needed those variables at run time for the Django app 🤔

File: ad-ops-dashboard/src/pages/Debug/Debug.tsx [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import SentryDebug from "../../components/Debug/SentryDebug";
+import * as Sentry from "@sentry/react";
+import { DefaultError } from "../../components/Errors/DefaultError";
+import Page404 from "../Page404";
+export default function Debug() {
+
+  if (process.env.NODE_ENV === "development") {
```
Comment (mashalifshin) [inline ad-ops-dashboard/src/pages/Debug/Debug.tsx:7]: This might be one of _those_ hot takes, so totally feel free to ignore! 

But I am not a fan of using `NODE_ENV` even for this. I know vite (and other frameworks) set this value and use it for stuff, but I think we should not use it for anything in our app code. It is hard to reason about all the things do `NODE_ENV` will do when an app has it used in various places as a toggle, and hard to reason about the multiple values it might have. It also makes it harder to spin off additional environments, like for example we have built this bundle for production but will run it in a new deployed dev or QA env, and we do want the debug menu to show...

So for this I prefer either using the existing `DEBUG` variable (if we want to think of the django env and the react env holistically here), or we can make one specific to just the react side...

**Review by mashalifshin — APPROVED** [review]
Thanks so much for the in depth conversations and for the attention to tests, this looks super nice now! The code coverage stats on the new components and overall are looking sooo much better after this last iteration. 

I replied on our conversations, but there is nothing merge blocking there.

I do think there might be some bugs with the current debug implementation, but I'm going to go ahead and Approve so you can get to merging right after you resolve those if needed.

Verified
- /debug page loads properly
- "Break useEffect" button logs a sentry event and shows DefaultError component

Possible issue with
-  "throw caught sentry error" button isn't doing anything currently
- "break component" button isn't logging a sentry error like the other button, but instead saying `Uncaught ReferenceError: ErrorComponent is not defined SentryDebug SentryDebug.tsx:40 ... SentryDebug.tsx:37`. However, it does seem to show the default error component here

File: ad-ops-dashboard/src/pages/Debug/Debug.tsx (hunk 7 above)
Comment (mashalifshin) [reply ad-ops-dashboard/src/pages/Debug/Debug.tsx:7]: (not sure if `DEBUG` is exactly the right choice, haven't dug deeply into it, but I think you get the gist of this particular hot take 😆)

**Comment by luc-lisi** [comment]
> Possible issue with
> 
>     * "throw caught sentry error" button isn't doing anything currently
> 
>     * "break component" button isn't logging a sentry error like the other button, but instead saying `Uncaught ReferenceError: ErrorComponent is not defined SentryDebug SentryDebug.tsx:40 ... SentryDebug.tsx:37`. However, it does seem to show the default error component here

Uh oh! Hmm, locally it seems to be working for me although I realize I forgot to re-update verification steps with the new env variables! In your `.env` file do you have the following variables set?
```
VITE_API_BASE_URL="http://127.0.0.1:7001/api/v1"
VITE_SENTRY_ENV="development"
VITE_SENTRY_DEBUG="false"
```

Otherwise if that doesn't fix it, it might be worth hopping on a call to verify what is going on! If you're running into this issue I'm sure someone else will too.

File: ad-ops-dashboard/src/pages/Debug/Debug.tsx (hunk 7 above)
Comment (luc-lisi) [reply ad-ops-dashboard/src/pages/Debug/Debug.tsx:7]: That is a hot take! I think I'm on board though with what you're saying though. 

I will add an optional `VITE_IS_DEBUG_PAGE_VISIBLE` env variable to make it super explicit exactly what behavior we expect 😄

File: ad-ops-dashboard/src/components/Debug/SentryDebug.test.tsx [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        element: (
+          <Sentry.ErrorBoundary fallback={DefaultError} showDialog>
+            <SentryDebug />
+          </Sentry.ErrorBoundary>
+        ),
+      }]
+    );
+  })
+
+  test('Sentry captureException is properly thrown when sentry error is raised', () => {
```
[the remark this answers is not in the document]
Comment (luc-lisi) [reply ad-ops-dashboard/src/components/Debug/SentryDebug.test.tsx:26]: Yes! This is a great call out. Fixed 😄

File: ad-ops-dashboard/src/pages/Debug/Debug.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (luc-lisi) [reply ad-ops-dashboard/src/pages/Debug/Debug.tsx:2]: Yeah agree'd! Definitely want to get that developer experience as good as possible 😄

File: ad-ops-dashboard/src/components/Errors/DefaultError.tsx (hunk 2 above)
Comment (luc-lisi) [reply ad-ops-dashboard/src/components/Errors/DefaultError.tsx:12]: Yeah it's a good call! Maybe I'll go with `DefaultErrorFallback` for clarity. That way it's super clear it's meant to be passed to the `fallback={...}` argument. Then for specific components we might want to wrap in their own boundaries can have names like `TableErrorFallback`, `FormErrorFallback`, etc.

File: ad-ops-dashboard/src/components/Debug/SentryDebug.test.tsx [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+      name: "Throw caught sentry error"
+    }));
+
+    const consoleOutput = consoleSpy.mock.calls.flat();
+
+    const hasSentryEvent = consoleOutput.some(message =>
+      typeof message === 'string' && message.includes("Sentry Event")
+    );
+
+    expect(hasSentryEvent).toBe(true);
```
[the remark this answers is not in the document]
Comment (luc-lisi) [reply ad-ops-dashboard/src/components/Debug/SentryDebug.test.tsx:42]: Oh, maybe I'm misunderstanding but for this test we are actually fully initializing sentry as defined in `instrument.tsx` in "is_local" mode and making sure a real Sentry Event gets emitted. So I think my goal here was to go a step further than just checking whether we call `captureException` but rather that `captureException` actually performs the behavior we expect (logging a sentry event to the console when is_local is true). 

This means that if `instrument.tsx` were edited in some way as to break sentry, this test would fail. I was actually worried this might be testing too many things at once 😅 

But let me know if I'm not quite catching what you're getting at! We could also add a spyOn here I think to additionally verify the call to captureException.

[the remark this answers is not in the document]
Comment (luc-lisi) [reply ad-ops-dashboard/src/components/Debug/SentryDebug.test.tsx:44]: Oh actually, adding an explicit check for that method call was super simple so I added it as well!

Let me know if this covers what you were getting at!

[the remark this answers is not in the document]
Comment (mashalifshin) [reply ad-ops-dashboard/src/components/Debug/SentryDebug.test.tsx:44]: That totally covers it, thank you! And thanks for explanation that this is the real sentry object, and totally makes sense we're testing the "isLocal" branch of the behavior here. So I think my comment isn't as salient given all that, but I still like that you asserted the method was called (also part of that branch of behavior). 

(I know it's considered brittle to assert on logs in specs, since usually that's not the actual behavior we care about verifying, but an additional side affect and one that is configurable based on environment at that. In this case the logging _is_ part of the actual behavior we want to assert, so I don't think this best practice applies here, but since it was fresh in my mind from another PR review, I thought I'd mention we should make sure the function gets called, too)

**Comment by mashalifshin** [comment]
> > Possible issue with
> > ```
> > * "throw caught sentry error" button isn't doing anything currently
> > 
> > * "break component" button isn't logging a sentry error like the other button, but instead saying `Uncaught ReferenceError: ErrorComponent is not defined SentryDebug SentryDebug.tsx:40 ... SentryDebug.tsx:37`. However, it does seem to show the default error component here
> > ```
> 
> @mashalifshin Uh oh! Hmm, locally it seems to be working for me although I realize I forgot to re-update verification steps with the new env variables! In your `.env` file do you have the following variables set?
> 
> ```
> VITE_API_BASE_URL="http://127.0.0.1:7001/api/v1"
> VITE_SENTRY_ENV="development"
> VITE_SENTRY_DEBUG="false"
> ```
> 
> Otherwise if that doesn't fix it, it might be worth hopping on a call to verify what is going on! If you're running into this issue I'm sure someone else will too.

I just tried again with actually npm installing the new dependency and the added environment variables, and everything is working as expected now! Thanks for bearing with me and let's :shipit:

File: ad-ops-dashboard/src/pages/Debug/Debug.tsx (hunk 7 above)
Comment (mashalifshin) [reply ad-ops-dashboard/src/pages/Debug/Debug.tsx]: Hahaha yeah that is really explicit. Thank you for your attention to this hot take. 😆

File: ad-ops-dashboard/src/components/Errors/DefaultError.tsx (hunk 2 above)
Comment (mashalifshin) [reply ad-ops-dashboard/src/components/Errors/DefaultError.tsx]: I think that turned out very nice and clear (the usage illustrated in the `Debug` page feels very clear/readable)

File: ad-ops-dashboard/src/components/Debug/SentryDebug.test.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (luc-lisi) [reply ad-ops-dashboard/src/components/Debug/SentryDebug.test.tsx:44]: Agree'd regarding asserting logs! Loved the suggestion to also add an explicit check for the captureException call as well so glad you appreciated it's inclusion 😄
